### PR TITLE
Define all conformance classes. Fix #80.

### LINKED ISSUE #80: Define conformance classes such as "DID document processor"
We should have a section that defines what is a "conforming DID document processor", "conforming DID document", etc. Those conformance classes would obviously require compliance with all normative statements in the specification, but also talk about producing errors, and about any other required behavior in certain situations.

This came up in https://github.com/w3c/did-spec/pull/66#discussion_r335776271

### PR DESCRIPTION
**Pull request by msporny**
This PR attempts to address issue #80 raised by @peacekeeper by adding the conformance classes contemplated by the specification. It is the first part of addressing @jricher's request in this week's DID WG meeting.

***
<a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/pull/191.html" title="Last updated on Feb 21, 2020, 2:45 AM UTC (7f46a0b)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/did-core/191/1b935d8...7f46a0b.html" title="Last updated on Feb 21, 2020, 2:45 AM UTC (7f46a0b)">Diff</a>

### CODE REVIEW

**Comment by jricher** [comment]
This is mostly good. I would suggest that we don't define terms like "conformant producer" and instead define "producer" as a specific role with actions and "conformance" separately in terms of the whole specification.

File: index.html [hunk 1]
```diff
@@ -635,7 +635,53 @@ <h2>Security and Privacy</h2>
...
+    </p>
+
+   <p>
+A <dfn>conforming DID</dfn> is any concrete expression of the rules
+specified in Section <a href="#identifier"></a> and MUST comply with relevant
+normative statements in that section.
+   </p>
+
+    <p>
+A <dfn>conforming DID Document</dfn> is any concrete expression of the data
```
Comment (peacekeeper) [inline index.html]:
```suggestion
A <dfn>conforming DID document</dfn> is any concrete expression of the data
```

File: index.html [hunk 2]
```diff
@@ -635,7 +635,53 @@ <h2>Security and Privacy</h2>
...
+model described in this specification and MUST comply with the relevant
+normative statements in Sections <a href="#data-model"></a> and  <a
+href="#core-properties"></a>. A serialization format for the conforming document
+MUST be deterministic, bi-directional, and lossless as described in Section <a
+href="#core-representations"></a>. The <a>conforming DID document</a> MAY be
+transmitted or stored in any such serialization format.
+    </p>
+
+    <p>
+A <dfn>conforming DID Method</dfn> is any specification that complies with the
```
Comment (peacekeeper) [inline index.html]:
```suggestion
A <dfn>conforming DID method</dfn> is any specification that complies with the
```

File: index.html [hunk 3]
```diff
@@ -635,7 +635,53 @@ <h2>Security and Privacy</h2>
...
+relevant normative statements in Section <a href="#methods"></a>.
+    </p>
+
+    <p>
+A <dfn>conforming producer</dfn> is any algorithm realized as software and/or
+hardware that generates a <a>conforming DID</a> or <a>conforming DID
+Document</a>. Conforming producers MUST NOT produce non-conforming
+<a>DIDs</a> or <a>DID Documents</a>.
```
Comment (peacekeeper) [inline index.html]:
```suggestion
<a>DIDs</a> or <a>DID documents</a>.
```

File: index.html [hunk 4]
```diff
@@ -635,7 +635,53 @@ <h2>Security and Privacy</h2>
...
+    </p>
+
+    <p>
+A <dfn>conforming consumer</dfn> is any algorithm realized as software and/or
+hardware that consumes a <a>conforming DID</a> or <a>conforming DID
+Document</a>. Conforming processors MUST produce errors when consuming
```
Comment (peacekeeper) [inline index.html]:
```suggestion
Document</a>. Conforming consumers MUST produce errors when consuming
```

File: index.html [hunk 5]
```diff
@@ -635,7 +635,53 @@ <h2>Security and Privacy</h2>
...
+   </p>
+
+    <p>
+A <dfn>conforming DID Document</dfn> is any concrete expression of the data
+model described in this specification and MUST comply with the relevant
+normative statements in Sections <a href="#data-model"></a> and  <a
+href="#core-properties"></a>. A serialization format for the conforming document
+MUST be deterministic, bi-directional, and lossless as described in Section <a
+href="#core-representations"></a>. The <a>conforming DID document</a> MAY be
+transmitted or stored in any such serialization format.
```
Comment (peacekeeper) [inline index.html]:
```suggestion
transmitted or stored in any such representation.
```

File: index.html [hunk 6]
```diff
@@ -635,7 +635,53 @@ <h2>Security and Privacy</h2>
...
+A <dfn>conforming DID</dfn> is any concrete expression of the rules
+specified in Section <a href="#identifier"></a> and MUST comply with relevant
+normative statements in that section.
+   </p>
+
+    <p>
+A <dfn>conforming DID Document</dfn> is any concrete expression of the data
+model described in this specification and MUST comply with the relevant
+normative statements in Sections <a href="#data-model"></a> and  <a
+href="#core-properties"></a>. A serialization format for the conforming document
```
Comment (peacekeeper) [inline index.html]:
```suggestion
href="#core-properties"></a>. A representation for the conforming document
```

**Comment by msporny** [comment]
> I would suggest that we don't define terms like "conformant producer" and instead define "producer" as a specific role with actions and "conformance" separately in terms of the whole specification.

Looking at how some other specs do this:

HTML5
https://html.spec.whatwg.org/multipage/infrastructure.html#conformance-classes

CSS
https://www.w3.org/TR/CSS/#css-levels

JSON-LD
https://w3c.github.io/json-ld-syntax/#conformance

No clear pattern, happy to go with what the group wants. I'll give @jricher's suggestion a go and see if it reads well. Anyone else have any strong opinions on this?

**Comment by dmitrizagidulin** [comment]
Fantastic work... 👍

**Comment by msporny** [comment]
> I'll give @jricher's suggestion a go and see if it reads well. 

I have applied @jricher's suggestion... it works, the language is a bit more awkward than before. I'm going to go with it and folks can put in other PRs if they have better language.

Normative, multiple reviews, changes applied, merging.

**Pull request merged**